File: apps/be-java-maven/src/main/java/com/soo/apps/webfilesystem/env/WebFileSystemBeEnv.java
package com.soo.apps.webfilesystem.env;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import jakarta.annotation.PreDestroy;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import com.soo.apps.webfilesystem.env.WebFileSystemBeEnvJson.WebFilesystem;

import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Component
@Getter
@Setter
public class WebFileSystemBeEnv extends com.soo.libs.env.Env<WebFileSystemBeEnvJson> {

    // ====================================================================================================
    // ====================================================================================================

    private String envJsonPathname;
    private String feEnvJsonPathname;

    private WebFileSystemFeEnvWatchService feEnvWatchService;

    @Value("${app.config.be-env-json}")
    public void setEnvJsonPathname(String envJsonPathname) {
        this.envJsonPathname = envJsonPathname;
        setFilePath(envJsonPathname);
        setClazz(WebFileSystemBeEnvJson.class);
        load(); // 부모 클래스에서 자동으로 파일 모니터링 시작
    }

    @PreDestroy
    public void cleanup() {
        stopFileWatching(); // 부모 클래스의 메서드 사용
        if (feEnvWatchService != null) {
            feEnvWatchService.stopWatchService();
        }
    }

    // ====================================================================================================
    // ====================================================================================================

    public WebFilesystem getWebFileSystem() {
        return getConfig() != null ? getConfig().getWebFilesystem() : null;
    }

    // ====================================================================================================

    public String getRootPath() {
        return getWebFileSystem() != null ? getWebFileSystem().getRootPath() : null;
    }

    public void setRootPath(String rootPath) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setRootPath(rootPath);
        }
    }

    // ====================================================================================================

    public String getUploadPath() {
        return getWebFileSystem() != null ? getWebFileSystem().getUploadPath() : null;
    }

    public void setUploadPath(String uploadPath) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setUploadPath(uploadPath);
        }
    }

    // ====================================================================================================

    public List<WebFileSystemBeEnvJson.AllowedRoot> getAllowedPaths() {
        return getWebFileSystem() != null ? getWebFileSystem().getAllowedPaths() : Collections.emptyList();
    }

    public List<String> getAllowedPathStrings() {
        return getWebFileSystem() != null ? getWebFileSystem().getAllowedPaths().stream()
                .map(WebFileSystemBeEnvJson.AllowedRoot::getPath)
                .collect(Collectors.toList()) : Collections.emptyList();
    }

    public void setAllowedPaths(List<WebFileSystemBeEnvJson.AllowedRoot> allowedPaths) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setAllowedPaths(allowedPaths);
        }
    }

    // ====================================================================================================

    public int getMaxDepth() {
        return getWebFileSystem() != null ? getWebFileSystem().getMaxDepth() : 0;
    }

    public void setMaxDepth(int maxDepth) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setMaxDepth(maxDepth);
        }
    }

    // ====================================================================================================

    public boolean isEnableHiddenFiles() {
        return getWebFileSystem() != null ? getWebFileSystem().isEnableHiddenFiles() : false;
    }

    public void setEnableHiddenFiles(boolean enableHiddenFiles) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setEnableHiddenFiles(enableHiddenFiles);
        }
    }

    // ====================================================================================================

    public boolean isShowJunctionPoints() {
        return getWebFileSystem() != null ? getWebFileSystem().isShowJunctionPoints() : false;
    }

    public void setShowJunctionPoints(boolean showJunctionPoints) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setShowJunctionPoints(showJunctionPoints);
        }
    }

    // ====================================================================================================

    public boolean isShowSymbolicLinks() {
        return getWebFileSystem() != null ? getWebFileSystem().isShowSymbolicLinks() : false;
    }

    public void setShowSymbolicLinks(boolean showSymbolicLinks) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setShowSymbolicLinks(showSymbolicLinks);
        }
    }

    // ====================================================================================================

    public boolean isHideInaccessibleLinks() {
        return getWebFileSystem() != null ? getWebFileSystem().isHideInaccessibleLinks() : false;
    }

    public void setHideInaccessibleLinks(boolean hideInaccessibleLinks) {
        if (getWebFileSystem() != null) {
            getWebFileSystem().setHideInaccessibleLinks(hideInaccessibleLinks);
        }
    }

    // ====================================================================================================
    // ====================================================================================================

}

File: apps/be-nestjs-ts/src/apps/webfilesystem/env/web.file.system.be.env.ts
import { Injectable, Logger, OnModuleDestroy } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as fs from 'fs';
import * as path from 'path';
import { WebFileSystemBeEnvJson, WebFilesystem, AllowedRoot } from './web.file.system.be.env.json';

@Injectable()
export class WebFileSystemBeEnv implements OnModuleDestroy {
    private readonly logger = new Logger(WebFileSystemBeEnv.name);
    private config: WebFileSystemBeEnvJson | null = null;
    private watcherActive: boolean = false;
    private envJsonPathname: string;

    constructor(private configService: ConfigService) {
        this.envJsonPathname = this.configService.get<string>('BE_ENV_JSON') || './config/env.json';
        this.load();
        this.startFileWatching();
    }

    onModuleDestroy() {
        this.stopFileWatching();
    }

    private load(): void {
        try {
            if (fs.existsSync(this.envJsonPathname)) {
                const jsonContent = fs.readFileSync(this.envJsonPathname, 'utf8');
                this.config = JSON.parse(jsonContent) as WebFileSystemBeEnvJson;
                this.logger.log(`Loaded WebFileSystem BE config from: ${this.envJsonPathname}`);
            } else {
                this.logger.warn(`Config file not found: ${this.envJsonPathname}`);
                this.config = { webFilesystem: this.getDefaultWebFilesystem() };
            }
        } catch (error) {
            this.logger.error(`Failed to load config from ${this.envJsonPathname}:`, error);
            this.config = { webFilesystem: this.getDefaultWebFilesystem() };
        }
    }

    private getDefaultWebFilesystem(): WebFilesystem {
        return {
            rootPath: '/tmp',
            uploadPath: '/tmp/uploads',
            allowedPaths: [],
            maxDepth: 5,
            enableHiddenFiles: false,
            showJunctionPoints: false,
            showSymbolicLinks: false,
            hideInaccessibleLinks: true
        };
    }

    private startFileWatching(): void {
        if (this.watcherActive) {
            this.stopFileWatching();
        }

        if (fs.existsSync(this.envJsonPathname)) {
            fs.watchFile(this.envJsonPathname, { interval: 1000 }, () => {
                this.logger.log(`Config file changed: ${this.envJsonPathname}`);
                this.load();
            });
            this.watcherActive = true;
            this.logger.log(`Started watching config file: ${this.envJsonPathname}`);
        }
    }

    private stopFileWatching(): void {
        if (this.watcherActive) {
            fs.unwatchFile(this.envJsonPathname);
            this.watcherActive = false;
            this.logger.log('Stopped watching config file');
        }
    }

    // ====================================================================================================
    // Getter methods for WebFilesystem config
    // ====================================================================================================

    getConfig(): WebFileSystemBeEnvJson | null {
        return this.config;
    }

    getWebFileSystem(): WebFilesystem | null {
        return this.config?.webFilesystem || null;
    }

    // ====================================================================================================
    // Root Path
    // ====================================================================================================

    getRootPath(): string | null {
        return this.getWebFileSystem()?.rootPath || null;
    }

    setRootPath(rootPath: string): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.rootPath = rootPath;
        }
    }

    // ====================================================================================================
    // Upload Path
    // ====================================================================================================

    getUploadPath(): string | null {
        return this.getWebFileSystem()?.uploadPath || null;
    }

    setUploadPath(uploadPath: string): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.uploadPath = uploadPath;
        }
    }

    // ====================================================================================================
    // Allowed Paths
    // ====================================================================================================

    getAllowedPaths(): AllowedRoot[] {
        return this.getWebFileSystem()?.allowedPaths || [];
    }

    getAllowedPathStrings(): string[] {
        return this.getAllowedPaths().map(root => root.path);
    }

    setAllowedPaths(allowedPaths: AllowedRoot[]): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.allowedPaths = allowedPaths;
        }
    }

    // ====================================================================================================
    // Max Depth
    // ====================================================================================================

    getMaxDepth(): number {
        return this.getWebFileSystem()?.maxDepth || 0;
    }

    setMaxDepth(maxDepth: number): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.maxDepth = maxDepth;
        }
    }

    // ====================================================================================================
    // Enable Hidden Files
    // ====================================================================================================

    isEnableHiddenFiles(): boolean {
        return this.getWebFileSystem()?.enableHiddenFiles || false;
    }

    setEnableHiddenFiles(enableHiddenFiles: boolean): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.enableHiddenFiles = enableHiddenFiles;
        }
    }

    // ====================================================================================================
    // Show Junction Points
    // ====================================================================================================

    isShowJunctionPoints(): boolean {
        return this.getWebFileSystem()?.showJunctionPoints || false;
    }

    setShowJunctionPoints(showJunctionPoints: boolean): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.showJunctionPoints = showJunctionPoints;
        }
    }

    // ====================================================================================================
    // Show Symbolic Links
    // ====================================================================================================

    isShowSymbolicLinks(): boolean {
        return this.getWebFileSystem()?.showSymbolicLinks || false;
    }

    setShowSymbolicLinks(showSymbolicLinks: boolean): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.showSymbolicLinks = showSymbolicLinks;
        }
    }

    // ====================================================================================================
    // Hide Inaccessible Links
    // ====================================================================================================

    isHideInaccessibleLinks(): boolean {
        return this.getWebFileSystem()?.hideInaccessibleLinks || false;
    }

    setHideInaccessibleLinks(hideInaccessibleLinks: boolean): void {
        const webFs = this.getWebFileSystem();
        if (webFs) {
            webFs.hideInaccessibleLinks = hideInaccessibleLinks;
        }
    }
}